File: slam/steps/colorize.mjs
// Порт colorize_laz (s20.py:1705-1893) + _load_calibration_camera
// (1637-1702). Реальный вызов в tasks.py (colorize.py:570:
// `s20.colorize_laz(local_in, cam_zip, frame_pose, local_out, cal_path)`)
// НИКОГДА не передаёт transform_path — та ветка (инверсия georeference-
// трансформа для уже-UTM облака) в реальном пайплайне мертвый код, потому
// что colorize всегда выполняется ДО georeference (см. georeference.mjs) —
// сознательно не портирована.
//
// ВАЖНО (перенесено как есть, не "исправлено"): несмотря на калибровочный
// ключ `fisheye_middle`, реальная проекция в исходном коде — обычная
// pinhole (fx,fy,cx,cy), БЕЗ компенсации дисторсии. Это поведение
// оригинала, не баг порта.
import { existsSync, statSync, mkdirSync, copyFileSync, readdirSync, readFileSync, rmSync } from "node:fs";
import { join, dirname } from "node:path";
import jpeg from "jpeg-js";
import yaml from "js-yaml";
import { runStep } from "../lib/stepContext.mjs";
import { readLas, writeLas } from "../lib/lasIO.mjs";
import { readFramePose } from "../lib/framePose.mjs";
import { extractZip } from "../lib/winZip.mjs";

function loadCalibrationCamera(calPath) {
  if (!calPath || !existsSync(calPath)) return null;
  try {
    let text = readFileSync(calPath, "utf8");
    text = text.replace(/^%YAML:[\d.]+\s*\n/, "").replaceAll("!!opencv-matrix", "");
    const d = yaml.load(text);
    if (!d) return null;
    const intrinsic = d.intrinsic || d;
    let cam = null;
    for (const key of ["fisheye_middle", "usb_cam", "nav_cam", "camera", "cam0"]) {
      if (intrinsic[key]) { cam = intrinsic[key]; break; }
    }
    if (!cam && intrinsic.camera_matrix) cam = intrinsic;
    if (!cam) return null;
    const kRaw = cam.camera_matrix || cam.K;
    if (!kRaw) return null;
    const kData = kRaw.data ? kRaw.data : (Array.isArray(kRaw[0]) ? kRaw.flat() : kRaw);
    if (kData.length < 9) return null;
    const fx = kData[0], fy = kData[4], cx = kData[2], cy = kData[5];

    let rCamInBody = [[1, 0, 0], [0, 1, 0], [0, 0, 1]];
    let tCamInBody = [0, 0, 0];
    const extRaw = d.extrinsic && d.extrinsic.lidar_middlecamera;
    if (extRaw) {
      const extData = extRaw.data ? extRaw.data : extRaw;
      if (extData.length >= 16) {
        // lidar_middlecamera maps LiDAR/body-frame -> camera frame; invert to
        // get camera pose in body frame (see s20.py _load_calibration_camera).
        const M = extData;
        const Rlc = [[M[0], M[1], M[2]], [M[4], M[5], M[6]], [M[8], M[9], M[10]]];
        const tlc = [M[3], M[7], M[11]];
        rCamInBody = transpose3(Rlc);
        tCamInBody = matVec3(scaleMat3(rCamInBody, -1), tlc);
      }
    }
    return { fx, fy, cx, cy, rCamInBody, tCamInBody };
  } catch {
    return null;
  }
}

function transpose3(m) { return [[m[0][0], m[1][0], m[2][0]], [m[0][1], m[1][1], m[2][1]], [m[0][2], m[1][2], m[2][2]]]; }
function scaleMat3(m, s) { return m.map((row) => row.map((v) => v * s)); }
function matVec3(m, v) { return [m[0][0] * v[0] + m[0][1] * v[1] + m[0][2] * v[2], m[1][0] * v[0] + m[1][1] * v[1] + m[1][2] * v[2], m[2][0] * v[0] + m[2][1] * v[1] + m[2][2] * v[2]]; }

function quatToMatrix(qx, qy, qz, qw) {
  return [
    [1 - 2 * (qy * qy + qz * qz), 2 * (qx * qy - qw * qz), 2 * (qx * qz + qw * qy)],
    [2 * (qx * qy + qw * qz), 1 - 2 * (qx * qx + qz * qz), 2 * (qy * qz - qw * qx)],
    [2 * (qx * qz - qw * qy), 2 * (qy * qz + qw * qx), 1 - 2 * (qx * qx + qy * qy)],
  ];
}

function interp1d(xs, ys, x) {
  const n = xs.length;
  if (x <= xs[0]) return ys[0];
  if (x >= xs[n - 1]) return ys[n - 1];
  let lo = 0, hi = n - 1;
  while (hi - lo > 1) { const mid = (lo + hi) >> 1; if (xs[mid] <= x) lo = mid; else hi = mid; }
  const t = (x - xs[lo]) / (xs[hi] - xs[lo]);
  return ys[lo] + t * (ys[hi] - ys[lo]);
}

function nearestFrameIndex(navTs, t) {
  // Port of np.searchsorted + "closer neighbour" pick in colorize_laz.
  let lo = 0, hi = navTs.length;
  while (lo < hi) { const mid = (lo + hi) >> 1; if (navTs[mid] < t) lo = mid + 1; else hi = mid; }
  const ni = Math.min(lo, navTs.length - 1);
  const li = Math.max(ni - 1, 0);
  return Math.abs(t - navTs[li]) < Math.abs(t - navTs[ni]) ? li : ni;
}

await runStep(async (ctx) => {
  const scanId = ctx.scan_id;
  const input = ctx.intermediate.filter_outliers;
  const output = ctx.intermediate.colorize;
  mkdirSync(dirname(output), { recursive: true });

  const inputs = new Map((ctx.inputs || []).map((i) => [i.kind, i.path]));
  if (!inputs.has("camera_frames")) {
    copyFileSync(input, output);
    return { assets: [{ type: "intermediate_laz", path: `${scanId}/intermediate/colorize.las`, size: statSync(output).size }] };
  }

  const cal = loadCalibrationCamera(inputs.get("calibration"));
  const { fx, fy, cx, cy, rCamInBody, tCamInBody } = cal || {
    fx: 500, fy: 500, cx: 320, cy: 240, rCamInBody: [[1, 0, 0], [0, 1, 0], [0, 0, 1]], tCamInBody: [0, 0, 0],
  };
  const rBodyInCam = transpose3(rCamInBody);

  const camExtractDir = join(ctx.scan_dir, "_camera_extract");
  if (existsSync(camExtractDir)) rmSync(camExtractDir, { recursive: true, force: true });
  extractZip(inputs.get("camera_frames"), camExtractDir);
  let navNames = readdirSync(camExtractDir).filter((n) => n.startsWith("nav_") && n.endsWith(".jpg")).sort();
  if (navNames.length === 0) navNames = readdirSync(camExtractDir).filter((n) => n.startsWith("left_") && n.endsWith(".jpg")).sort();
  if (navNames.length === 0) {
    copyFileSync(input, output);
    rmSync(camExtractDir, { recursive: true, force: true });
    return { assets: [{ type: "intermediate_laz", path: `${scanId}/intermediate/colorize.las`, size: statSync(output).size }] };
  }
  const navTs = navNames.map((n) => Number(n.split("_")[1].split(".")[0]) / 1e9);

  const framePose = readFramePose(inputs.get("frame_pose"));
  // Continuous-sign quaternion (same fix-up as framePose.mjs's interpFramePose).
  const qx = Float64Array.from(framePose.qx), qy = Float64Array.from(framePose.qy);
  const qz = Float64Array.from(framePose.qz), qw = Float64Array.from(framePose.qw);
  for (let i = 1; i < qx.length; i++) {
    const dot = qx[i] * qx[i - 1] + qy[i] * qy[i - 1] + qz[i] * qz[i - 1] + qw[i] * qw[i - 1];
    if (dot < 0) { qx[i] *= -1; qy[i] *= -1; qz[i] *= -1; qw[i] *= -1; }
  }
  const tMin = framePose.times[0], tMax = framePose.times[framePose.times.length - 1];

  const las = readLas(input);
  const n = las.count;
  const ptTs = new Float64Array(n);
  const hasGpsTime = las.gpsTime && las.gpsTime.some((v) => v !== 0);
  for (let i = 0; i < n; i++) ptTs[i] = hasGpsTime ? Math.min(Math.max(las.gpsTime[i], tMin), tMax) : tMin + ((tMax - tMin) * i) / Math.max(n - 1, 1);

  // Bucket point indices by nearest camera frame (mirrors the vectorised
  // `ni == fi` mask loop in s20.py without materialising a full mask array).
  const buckets = new Map();
  for (let i = 0; i < n; i++) {
    const fi = nearestFrameIndex(navTs, ptTs[i]);
    if (!buckets.has(fi)) buckets.set(fi, []);
    buckets.get(fi).push(i);
  }

  const red = new Uint8Array(n), green = new Uint8Array(n), blue = new Uint8Array(n);
  for (const [fi, idxs] of buckets) {
    let img;
    try {
      const raw = readFileSync(join(camExtractDir, navNames[fi]));
      img = jpeg.decode(raw, { useTArray: true });
    } catch {
      continue;
    }
    const tf = Math.min(Math.max(navTs[fi], tMin), tMax);
    const itx = interp1d(framePose.times, framePose.tx, tf);
    const ity = interp1d(framePose.times, framePose.ty, tf);
    const itz = interp1d(framePose.times, framePose.tz, tf);
    let iqx = interp1d(framePose.times, qx, tf), iqy = interp1d(framePose.times, qy, tf);
    let iqz = interp1d(framePose.times, qz, tf), iqw = interp1d(framePose.times, qw, tf);
    const nr = Math.hypot(iqx, iqy, iqz, iqw);
    if (nr < 1e-9) continue;
    iqx /= nr; iqy /= nr; iqz /= nr; iqw /= nr;
    const rBw = quatToMatrix(iqx, iqy, iqz, iqw);
    const rWb = transpose3(rBw);

    for (const i of idxs) {
      const pw = [las.x[i] - itx, las.y[i] - ity, las.z[i] - itz];
      const pBody = matVec3(rWb, pw);
      const pCam = matVec3(rBodyInCam, [pBody[0] - tCamInBody[0], pBody[1] - tCamInBody[1], pBody[2] - tCamInBody[2]]);
      if (pCam[2] <= 0.1) continue;
      const u = Math.round((fx * pCam[0]) / pCam[2] + cx);
      const v = Math.round((fy * pCam[1]) / pCam[2] + cy);
      if (u < 0 || u >= img.width || v < 0 || v >= img.height) continue;
      const off = (v * img.width + u) * 4; // jpeg-js decodes to RGBA
      red[i] = img.data[off]; green[i] = img.data[off + 1]; blue[i] = img.data[off + 2];
    }
  }

  writeLas(output, {
    pointFormat: 7, x: las.x, y: las.y, z: las.z,
    gpsTime: las.gpsTime, intensity: las.intensity, red, green, blue,
  });
  rmSync(camExtractDir, { recursive: true, force: true });

  return { assets: [{ type: "intermediate_laz", path: `${scanId}/intermediate/colorize.las`, size: statSync(output).size }] };
});

File: pyserver/apps/slam/pipeline/steps/colorize.py
"""
Port of colorize_laz (s20.py:1705-1893) + _load_calibration_camera
(1637-1702). The real call site in tasks.py (colorize.py:570:
`s20.colorize_laz(local_in, cam_zip, frame_pose, local_out, cal_path)`)
NEVER passes transform_path -- that branch (inverting the georeference
transform for an already-UTM cloud) is dead code in the real pipeline,
because colorize always runs BEFORE georeference (see georeference.py) --
deliberately not ported.

IMPORTANT (carried over as-is, not "fixed"): despite the calibration key
`fisheye_middle`, the real projection in the source is plain pinhole
(fx,fy,cx,cy), WITHOUT distortion compensation. That's the original's
behaviour, not a porting bug.
"""

import re
import shutil
import zipfile
from pathlib import Path

import numpy as np
import yaml
from PIL import Image

from ..frame_pose import read_frame_pose
from ..las_io import read_las, write_las


def _load_calibration_camera(cal_path):
    if not cal_path or not Path(cal_path).is_file():
        return None
    try:
        text = Path(cal_path).read_text(encoding="utf-8")
        text = re.sub(r"^%YAML:[\d.]+\s*\n", "", text)
        text = text.replace("!!opencv-matrix", "")
        d = yaml.safe_load(text)
        if not d:
            return None
        intrinsic = d.get("intrinsic", d)
        cam = None
        for key in ("fisheye_middle", "usb_cam", "nav_cam", "camera", "cam0"):
            if intrinsic.get(key):
                cam = intrinsic[key]
                break
        if not cam and intrinsic.get("camera_matrix"):
            cam = intrinsic
        if not cam:
            return None
        k_raw = cam.get("camera_matrix") or cam.get("K")
        if not k_raw:
            return None
        if isinstance(k_raw, dict) and "data" in k_raw:
            k_data = k_raw["data"]
        elif isinstance(k_raw, list) and k_raw and isinstance(k_raw[0], list):
            k_data = [v for row in k_raw for v in row]
        else:
            k_data = k_raw
        if len(k_data) < 9:
            return None
        fx, fy, cx, cy = k_data[0], k_data[4], k_data[2], k_data[5]

        r_cam_in_body = np.eye(3)
        t_cam_in_body = np.zeros(3)
        ext_raw = (d.get("extrinsic") or {}).get("lidar_middlecamera")
        if ext_raw:
            ext_data = ext_raw["data"] if isinstance(ext_raw, dict) and "data" in ext_raw else ext_raw
            if len(ext_data) >= 16:
                # lidar_middlecamera maps LiDAR/body-frame -> camera frame;
                # invert to get camera pose in body frame (see
                # s20.py _load_calibration_camera).
                m = ext_data
                r_lc = np.array([[m[0], m[1], m[2]], [m[4], m[5], m[6]], [m[8], m[9], m[10]]])
                t_lc = np.array([m[3], m[7], m[11]])
                r_cam_in_body = r_lc.T
                t_cam_in_body = -r_cam_in_body @ t_lc
        return {"fx": fx, "fy": fy, "cx": cx, "cy": cy, "r_cam_in_body": r_cam_in_body, "t_cam_in_body": t_cam_in_body}
    except (OSError, yaml.YAMLError, KeyError, IndexError, TypeError):
        return None


def _quat_to_matrix(qx, qy, qz, qw):
    return np.array([
        [1 - 2 * (qy * qy + qz * qz), 2 * (qx * qy - qw * qz), 2 * (qx * qz + qw * qy)],
        [2 * (qx * qy + qw * qz), 1 - 2 * (qx * qx + qz * qz), 2 * (qy * qz - qw * qx)],
        [2 * (qx * qz - qw * qy), 2 * (qy * qz + qw * qx), 1 - 2 * (qx * qx + qy * qy)],
    ])


def run(ctx: dict) -> dict:
    scan_id = ctx["scan_id"]
    scan_dir = Path(ctx["scan_dir"])
    input_path = ctx["intermediate"]["filter_outliers"]
    output_path = ctx["intermediate"]["colorize"]
    Path(output_path).parent.mkdir(parents=True, exist_ok=True)

    inputs = {i["kind"]: i["path"] for i in (ctx.get("inputs") or [])}
    if "camera_frames" not in inputs:
        shutil.copyfile(input_path, output_path)
        return {"assets": [{"type": "intermediate_laz", "path": f"{scan_id}/intermediate/colorize.las", "size": Path(output_path).stat().st_size}]}

    cal = _load_calibration_camera(inputs.get("calibration"))
    if cal:
        fx, fy, cx, cy = cal["fx"], cal["fy"], cal["cx"], cal["cy"]
        r_cam_in_body, t_cam_in_body = cal["r_cam_in_body"], cal["t_cam_in_body"]
    else:
        fx, fy, cx, cy = 500, 500, 320, 240
        r_cam_in_body, t_cam_in_body = np.eye(3), np.zeros(3)
    r_body_in_cam = r_cam_in_body.T

    cam_extract_dir = scan_dir / "_camera_extract"
    if cam_extract_dir.exists():
        shutil.rmtree(cam_extract_dir)
    with zipfile.ZipFile(inputs["camera_frames"]) as zf:
        zf.extractall(cam_extract_dir)

    nav_names = sorted(p.name for p in cam_extract_dir.iterdir() if p.name.startswith("nav_") and p.name.endswith(".jpg"))
    if not nav_names:
        nav_names = sorted(p.name for p in cam_extract_dir.iterdir() if p.name.startswith("left_") and p.name.endswith(".jpg"))
    if not nav_names:
        shutil.copyfile(input_path, output_path)
        shutil.rmtree(cam_extract_dir)
        return {"assets": [{"type": "intermediate_laz", "path": f"{scan_id}/intermediate/colorize.las", "size": Path(output_path).stat().st_size}]}
    nav_ts = np.array([int(n.split("_")[1].split(".")[0]) / 1e9 for n in nav_names])

    frame_pose = read_frame_pose(inputs["frame_pose"])
    # Continuous-sign quaternion (same fix-up as frame_pose.py's interp_frame_pose).
    qx, qy, qz, qw = frame_pose.qx.copy(), frame_pose.qy.copy(), frame_pose.qz.copy(), frame_pose.qw.copy()
    dot = qx[1:] * qx[:-1] + qy[1:] * qy[:-1] + qz[1:] * qz[:-1] + qw[1:] * qw[:-1]
    flip = np.cumprod(np.where(dot < 0, -1.0, 1.0))
    qx[1:] *= flip
    qy[1:] *= flip
    qz[1:] *= flip
    qw[1:] *= flip
    t_min, t_max = frame_pose.times[0], frame_pose.times[-1]

    las = read_las(input_path)
    n = las.count
    has_gps_time = las.gps_time is not None and np.any(las.gps_time != 0)
    if has_gps_time:
        pt_ts = np.clip(las.gps_time, t_min, t_max)
    else:
        pt_ts = t_min + (t_max - t_min) * np.arange(n) / max(n - 1, 1)

    # Bucket point indices by nearest camera frame (np.searchsorted +
    # "closer neighbour" pick, mirrors nearestFrameIndex in the .mjs port).
    ni = np.searchsorted(nav_ts, pt_ts)
    ni = np.clip(ni, 0, len(nav_ts) - 1)
    li = np.clip(ni - 1, 0, None)
    closer_to_li = np.abs(pt_ts - nav_ts[li]) < np.abs(pt_ts - nav_ts[ni])
    frame_idx = np.where(closer_to_li, li, ni)

    red = np.zeros(n, dtype=np.uint8)
    green = np.zeros(n, dtype=np.uint8)
    blue = np.zeros(n, dtype=np.uint8)

    for fi in np.unique(frame_idx):
        idxs = np.nonzero(frame_idx == fi)[0]
        try:
            img = Image.open(cam_extract_dir / nav_names[fi]).convert("RGB")
        except (OSError, ValueError):
            continue
        img_arr = np.asarray(img)
        h, w = img_arr.shape[0], img_arr.shape[1]

        tf = min(max(nav_ts[fi], t_min), t_max)
        itx = np.interp(tf, frame_pose.times, frame_pose.tx)
        ity = np.interp(tf, frame_pose.times, frame_pose.ty)
        itz = np.interp(tf, frame_pose.times, frame_pose.tz)
        iqx = np.interp(tf, frame_pose.times, qx)
        iqy = np.interp(tf, frame_pose.times, qy)
        iqz = np.interp(tf, frame_pose.times, qz)
        iqw = np.interp(tf, frame_pose.times, qw)
        nr = np.hypot(np.hypot(iqx, iqy), np.hypot(iqz, iqw))
        if nr < 1e-9:
            continue
        iqx, iqy, iqz, iqw = iqx / nr, iqy / nr, iqz / nr, iqw / nr
        r_bw = _quat_to_matrix(iqx, iqy, iqz, iqw)
        r_wb = r_bw.T

        pw = np.stack([las.x[idxs] - itx, las.y[idxs] - ity, las.z[idxs] - itz], axis=1)
        p_body = pw @ r_wb.T
        p_cam = (p_body - t_cam_in_body) @ r_body_in_cam.T

        valid = p_cam[:, 2] > 0.1
        u = np.round((fx * p_cam[:, 0]) / np.where(valid, p_cam[:, 2], 1) + cx).astype(np.int64)
        v = np.round((fy * p_cam[:, 1]) / np.where(valid, p_cam[:, 2], 1) + cy).astype(np.int64)
        in_bounds = valid & (u >= 0) & (u < w) & (v >= 0) & (v < h)

        sel = idxs[in_bounds]
        su, sv = u[in_bounds], v[in_bounds]
        red[sel] = img_arr[sv, su, 0]
        green[sel] = img_arr[sv, su, 1]
        blue[sel] = img_arr[sv, su, 2]

    write_las(
        output_path, point_format=7, x=las.x, y=las.y, z=las.z,
        gps_time=las.gps_time, intensity=las.intensity, red=red, green=green, blue=blue,
    )
    shutil.rmtree(cam_extract_dir)

    return {"assets": [{"type": "intermediate_laz", "path": f"{scan_id}/intermediate/colorize.las", "size": Path(output_path).stat().st_size}]}

